Sources/NetworkProtection/Status/KnownFailure.swift

```diff
@@ -0,0 +1,50 @@
+//
+//  KnownFailure.swift
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+@objc
+final public class KnownFailure: NSObject, Codable {
```

Comment:
This forces all errors that are "known" to be defined in BSK which seems unnecessarily restrictive.

Why instead not just make the error adhere to `Codable` and store that in `NetworkProtectionKnownFailureStore`.

Then the VPN app can decide what errors are worth passing over IPC, but at least we don't have to add errors that are macOS specific to BSK.

Reply:
> Why instead not just make the error adhere to Codable and store that in NetworkProtectionKnownFailureStore.

It's tricky as we need a concrete type to decode the error; and that type needs to be passed around in different ways (e.g. as part of `XPCClientInterface`, distributed notification object).

To keep it simple, I've moved the custom handling code for `TunnelError` outside of the initializer and instead made it so that an error conforming to `InternalErrorWrapping` can expose its internal error. That way we can keep specific errors outside of BSK.

Reply:
Understood.

That said, I do think that instead of encoding domain, code and localizedDescription, we should turn this into a `Codable` enum.  See [this comment for context](https://github.com/duckduckgo/macos-browser/pull/2773#discussion_r1617249489).